core/src/main/scala/chisel3/IO.scala

```diff
@@ -44,3 +44,42 @@ object IO {
     iodefClone
   }
 }
+
+private[chisel3] object SecretIO {
+
+  /** Constructs a secret port for the current Module
```

Comment:
is a SecretIO visible in DataMirror getPorts and friends? if not why not?

Reply:
It isn't, and that's to avoid the issue of mutability of a definition. To expand further, if I have code like:
```
val child = Module(new Child)
...
functionFoo(DataMirror.getPorts(child))
...
val myWire = BoringUtils.bore(child.wire)
...
functionBar(DataMirror.getPorts(child))
```

`functionFoo` and `functionBar` would get different ports, which would be difficult to debug because the order of execution matters. This is in general why Chisel has a module "close", so that once I have a reference to `child`, all its Chisel-y things are known.

Another reason supporting not making these present in DataMirror APIs is that all existing uses of boring utils would never show up in a DataMirror getPorts because they were added by CIRCT, so this is backwards compatible behavior.

Reply:
sure, but is this a silent failure of the DataMirror APIs that they would return something different than the user might expect?

Reply:
I can update their documentation to report that they return all non-secret ports?

Reply:
Another way we could do this is if `DataMirror.getPorts` calls `.toDefinition` to super-close a module prior to returning all the ports? That way you still get all the ports, but you never get a partial list (e.g. if someone adds ports later, it will error)?

Reply:
Added clear error messages, good discussion, thanks!